### PR TITLE
Autocomplete: Allow user to scroll through items in the dropdown

### PR DESCRIPTION
**Pull request by mltejada**
## Description

References https://github.com/trimble-oss/modus-web-components/issues/2081

https://e-builderinc.atlassian.net/browse/EPD-11996

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.tsx [hunk 1]
```diff
@@ -290,7 +330,9 @@ export class ModusAutocomplete {
           if (this.hasFocus) {
             this.hasFocus = this.disableCloseOnSelect;
           }
-        }}>
+        }}
+        onKeyDown={(e) => this.handleInputKeyDown(e)}
+      >
```
Comment (cjwinsor) [inline stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.tsx:335]: This formatting would be changed when `npm run lint` is run.  Can you run that command and ensure all lint formatting is standardized?   It'll move the `>` back to the end of the line.

File: stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.tsx [hunk 2]
```diff
@@ -302,8 +344,10 @@ export class ModusAutocomplete {
                   <li
                     class="text-option"
                     tabindex="0"
+                    role="option"
```
Comment (cjwinsor) [inline stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.tsx:347]: I think this was just added in another PR which may be one of the sources of conflict with the main branch.  Try bringing in main to resolve.

File: stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.tsx [hunk 3]
```diff
@@ -168,15 +169,40 @@ export class ModusAutocomplete {
     this.hasFocus = !this.disableCloseOnSelect;
   };
 
-  handleOptionKeyPress = (event: any, option: any, isCustomOption = false) => {
-    this.disableFiltering = !this.disableCloseOnSelect;
-    if (event.key !== 'Enter') {
-      return;
+  handleInputKeyDown = (event: KeyboardEvent) => {
+    if (!this.hasFocus || event.defaultPrevented || event.altKey) {
```
Comment (cjwinsor) [inline stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.tsx:173]: I'm curious the use case for the `alt` key here.  What is the scenario being avoided?

**Review by cjwinsor — CHANGES_REQUESTED** [review]
There is an issue when mixing tabs and arrow keys due to tabbing not updating the focus index.   For the example of Apples, Banana, Orange: 

- Type a then tab 3 times to highlight Orange
- Press the down arrow once
- Banana gets highlighted since the focus index incremented from 0 to 1

File: stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.tsx (hunk 1 above)
Comment (mltejada) [reply stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.tsx]: Done. I'll have to check my settings as every time I save it auto formats to this.

File: stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.tsx (hunk 3 above)
Comment (mltejada) [reply stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.tsx]: I forgot to remove this. Validated is not needed .

**Comment by cjwinsor** [comment]
@mltejada We still need to support Tabs in addition to Up an Down.

**Comment by mltejada** [comment]
> @mltejada We still need to support Tabs in addition to Up and Down.

I meet with Carlos and confirmed this is the expected behavior from https://modus.ideas.aha.io/ideas/MOD-I-14

Note: We have another story to support tab on the "X" icon but tab won't be used for the dropdown options.

**Comment by cjwinsor** [comment]
> > @mltejada We still need to support Tabs in addition to Up and Down.
> 
> I meet with Carlos and confirmed this is the expected behavior from https://modus.ideas.aha.io/ideas/MOD-I-14
> 
> The tab will move focus to the next component, but will not scroll down or up through the options.
> 
> Note: We have another story to support tab on the "X" icon (currently not working).

Let me check with the team.  Upon review of at least two other component library I see it functioning as you describe.  The modus submission did not make clear that tabbing through the items would be removed (In the approved writeup).

**Comment by mltejada** [comment]
> Let me check with the team. Upon review of at least two other component library I see it functioning as you describe. The modus submission did not make clear that tabbing through the items would be removed (In the approved writeup).

Thanks Chris. Let me know. I didn't originally work on the tab because we have another story for it, but since it was conflicting then I confirmed the behavior with Carlos. 
Note: I added tests for the custom options scenario.

File: stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.e2e.ts [hunk 4]
```diff
@@ -459,4 +459,214 @@ describe('modus-autocomplete', () => {
...
+
+    // Press Down Arrow
+    await page.keyboard.press('ArrowDown');
+    await page.waitForChanges();
+
+    await page.keyboard.press('Enter');
+    await page.waitForChanges();
+
+    expect(optionSelected).toHaveReceivedEvent();
+    expect(optionSelected).toHaveReceivedEventDetail("1");
```
Comment (cjwinsor) [inline stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.e2e.ts:590]: @mltejada Use single quotes

File: stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.e2e.ts [hunk 5]
```diff
@@ -459,4 +459,214 @@ describe('modus-autocomplete', () => {
...
+
+    // Press Down Arrow
+    await page.keyboard.press('ArrowDown');
+    await page.waitForChanges();
+
+    await page.keyboard.press('Enter');
+    await page.waitForChanges();
+
+    expect(optionSelected).toHaveReceivedEvent();
+    expect(optionSelected).toHaveReceivedEventDetail("2");
```
Comment (cjwinsor) [inline stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.e2e.ts:628]: @mltejada Use single quotes

File: stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.e2e.ts [hunk 6]
```diff
@@ -459,4 +459,214 @@ describe('modus-autocomplete', () => {
...
+
+    // Press Down Up
+    await page.keyboard.press('ArrowUp');
+    await page.waitForChanges();
+
+    await page.keyboard.press('Enter');
+    await page.waitForChanges();
+
+    expect(optionSelected).toHaveReceivedEvent();
+    expect(optionSelected).toHaveReceivedEventDetail("1");
```
Comment (cjwinsor) [inline stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.e2e.ts:670]: @mltejada Use single quotes

**Review by cjwinsor — CHANGES_REQUESTED** [review]

File: stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.e2e.ts (hunk 4 above)
Comment (mltejada) [reply stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.e2e.ts]: Updated

File: stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.e2e.ts (hunk 5 above)
Comment (mltejada) [reply stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.e2e.ts]: Updated

File: stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.e2e.ts (hunk 6 above)
Comment (mltejada) [reply stencil-workspace/src/components/modus-autocomplete/modus-autocomplete.e2e.ts]: Updated

**Review by cjwinsor — APPROVED** [review]

**Pull request merged**